Make all URLs absolute to support proxies

quinox opened the pull request:
Related:  #528 and #519

**Feedback greatly appreciated**

This code implement a configuration option to set a global URL prefix, making all links start with the given prefix. The default prefix is `/` (so nothing changes). 

### Rationale ###

I would like to run Prometheus (and PromDash eventually) behind a proxy. This proxy does the user authentication, and once approved proxies requests for certain prefixes (`/graphite/`, `/jenkins/`, `/smokeping/` etc) to services on separate machines. When these services use the same prefix on their own (fe. `http://jenkins-vm/jenkins/`) you can pass responses from these services back to the user unaltered; otherwise you'd have to let the proxy rewrite links from a response (`<a href="/build/3/">`) to include the prefix (`<a href="/jenkins/build/3/">`) to make sure the user can click on things.

### Alternatives ###

Using a separate DNS entry (as suggested in #528) would not be nice because then users would have to re-enter their credentials when using Prometheus (twice if they also want to use PromDash). 

## ToDo ##

* Needs to be code reviewed, I've never worked with golang nor the Prometheus codebase before
* ~~Code under `notifications/` and `rules/` have to use the URL prefix as well somehow~~
* ~~Prometheus should do some sanity checking on the setting to make sure the user starts and ends the option with a slash~~
* ~~Prometheus should do some escaping on the setting, otherwise a prefix of `/foo">/` will break all templates~~

juliusv commented on web/web.go:
```diff
@@ -57,42 +58,48 @@ func (ws WebService) ServeForever() error {
+		http.Handle(*urlPrefix + "-/quit", http.HandlerFunc(ws.quitHandler))
+	}
+
+	if *urlPrefix != "/" {
```
Ah, I guess this is a neat way of supporting both proxy configurations which drop the prefix as well as ones that don't?

juliusv commented on web/web.go:
```diff
@@ -130,22 +137,27 @@ func getTemplateFile(name string) (string, error) {
 	t.Funcs(template.FuncMap{
 		"since":       time.Since,
 		"getConsoles": getConsoles,
+		"getUrlPrefix": getUrlPrefix,
```
In Go, getters are usually not called `getThing`, just `thing` (perhaps `getConsoles` should also be changed then, though you could argue that it's not just a getter, but actually does quite some stuff - up for review). Anyways, I'd just call this `urlPrefix` (you might need to call the Go function differently though since `urlPrefix` is already a variable).

juliusv commented on web/templates/_base.html:
```diff
@@ -3,10 +3,12 @@
+    <link type="text/css" rel="stylesheet" href="{{ getUrlPrefix }}static/vendor/bootstrap-3.3.1/css/bootstrap.min.css">
+    <link type="text/css" rel="stylesheet" href="{{ getUrlPrefix }}static/css/prometheus.css">
+
+    <script>URL_PREFIX = "{{ getUrlPrefix }}";</script>
```
`var URL_PREFIX` please

juliusv commented on web/static/js/graph_template.handlebar:
```diff
@@ -1,5 +1,5 @@
         <div id="graph_wrapper{{id}}" class="graph_wrapper">
-          <form action="/api/query_range" method="GET" class="query_form form-inline">
+          <form action="{{ getUrlPrefix }}api/query_range" method="GET" class="query_form form-inline">
```
Shouldn't this be `URL_PREFIX` too, since it's not a Go template?

Did you try graphing stuff, and does it still work? Actually I'd have to check how handlebar templates work and whether these `{{ ... }}` expressions may include any JS variable, or only the ones that were explicitly passed to the template.

juliusv commented:
Thanks! Will give this a closer look over the next days, and hope others will chime in too.

quinox commented:
Thanks, I'll process your feedback tonight.

This changeset does allow you to open the interface, click on all links and make graphs. As far I could tell everything works, but undoubtly I have missed some stuff. The consoles-part I fixed blindly since I don't have any yet.

The situation of no-prefix and some-prefix are treated the same so you don't need any special checks / if-statements throughout the codebase. The only exception is the redirect from `/` to `PREFIX` in case your prefix isn't the same as `/`, otherwise you'd get a 404 when visiting `http://host/`.

quinox commented:
* The Handlebars template was indeed broken; I fixed it by adding a Handlebars helper called `urlPrefix`
* There was no need to escape urlPrefix: apparently golang is context aware when working with templates and does the right thing everywhere (neat!)
* The URL prefix is now checked and corrected inside `web/web.go:ServeForever`, but this is really too late: The sanitized version should already be accessible inside `main.go` when fe. the `NotificationHandler` is created. I started moving the flag into `main.go` but it turned into a horrible mess.

quinox commented:
I've discussed it with @juliusv on IRC the other day: In order to make `executeTemplate`, `getTemplate`, `getConsoles` etc. aware of the prefix we should make some kind of `templateExpander` struct which holds those methods. Then `main.go` can pass the prefix string to the separate structs, each struct should then create its own `templateExpander`.

juliusv commented on web/templates/_base.html:
```diff
@@ -3,10 +3,12 @@
+    <link type="text/css" rel="stylesheet" href="{{ pathPrefix }}static/vendor/bootstrap-3.3.1/css/bootstrap.min.css">
+    <link type="text/css" rel="stylesheet" href="{{ pathPrefix }}static/css/prometheus.css">
+
+    <script>var PATH_PREFIX = "{{ pathPrefix }}";</script>
```
This line will also be needed in `console_libraries/prom.lib`, as consoles use that as their base template. (tested and it works when adding it there)

juliusv commented on main.go:
```diff
@@ -65,6 +66,8 @@ var (
 	storageDirty          = flag.Bool("storage.local.dirty", false, "If set, the local storage layer will perform crash recovery even if the last shutdown appears to be clean.")
 	storagePedanticChecks = flag.Bool("storage.local.pedantic-checks", false, "If set, a crash recovery will perform checks on each series file. This might take a very long time.")
 
+	pathPrefix = flag.String("web.path-prefix", "/", "Prefix for all paths, defaults to /")
```
The default will be displayed automatically in the help output, so I'd omit that and just write `Prefix for all web paths.`.

juliusv commented on web/alerts.go:
```diff
@@ -50,6 +50,8 @@ type AlertsHandler struct {
 
 	mutex sync.Mutex
+
+	PathPrefix string
```
The empty lines here are getting weird :) Let's get rid of them completely and put the mutex at the very top.

juliusv commented on web/api/api.go:
```diff
@@ -31,19 +31,19 @@ type MetricsService struct {
 
 // RegisterHandler registers the handler for the various endpoints below /api.
-func (msrv *MetricsService) RegisterHandler() {
+func (msrv *MetricsService) RegisterHandler(pathPrefix *string) {
```
It's uncommon to pass a string as a pointer, as they are essentially passed by reference anyways (and immutable). So just `string` here.

juliusv commented on web/web.go:
```diff
@@ -47,52 +47,60 @@ type WebService struct {
 
 // ServeForever serves the HTTP endpoints and only returns upon errors.
-func (ws WebService) ServeForever() error {
+func (ws WebService) ServeForever(pathPrefix *string) error {
```
Dito here for the non-pointer string type.

juliusv commented on web/web.go:
```diff
@@ -188,7 +198,7 @@ func dumpHeap(w http.ResponseWriter, r *http.Request) {
 
 // MustBuildServerURL returns the server URL and panics in case an error occurs.
-func MustBuildServerURL() string {
+func MustBuildServerURL(pathPrefix *string) string {
```
Dito.

juliusv commented:
:+1: Besides the comments.

Confirmed that everything is working, including links back from alertmanager to Prometheus with a prefix, and consoles also work once the one-line fix from my comment is applied. Thanks a lot!

brian-brazil commented on web/static/js/prom_console.js:
```diff
@@ -502,7 +502,7 @@ PromConsole.Graph.prototype.dispatch = function() {
   for (var i = 0; i < this.params.expr.length; ++i) {
     var endTime = this.params.endTime;
-    var url = "/api/query_range?expr=" + encodeURIComponent(this.params.expr[i])
+    var url = PATH_PREFIX + "api/query_range?expr=" + encodeURIComponent(this.params.expr[i])
```
Where are you getting PATH_PREFIX from here?

I think you need to add it in prom.lib

juliusv replied:
Yep, see this comment: https://github.com/prometheus/prometheus/pull/614#discussion_r27608073

brian-brazil commented on web/web.go:
```diff
@@ -35,7 +35,7 @@ import (
 var (
 	listenAddress  = flag.String("web.listen-address", ":9090", "Address to listen on for the web interface, API, and telemetry.")
-	metricsPath    = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics.")
+	metricsPath    = flag.String("web.telemetry-path", "metrics", "Path under which to expose metrics.")
```
This change isn't consistent with the other servers such as node_exporter. We probably need to sanatize this.

juliusv replied:
Good point. Yes, this should still start with a slash to be consistent with the other servers.

quinox commented:
Thanks, I've incorporated all remarks.

brian-brazil commented:
:+1: 

You seem to have a built failure.

juliusv commented:
:+1: Awesome! Could you squash, then I'll merge.

quinox commented:
Squashed.

juliusv commented:
Thanks a lot!
